k2/addons/k2-serial/zserial/connection.py

```diff
@@ -27,29 +27,48 @@ class SerialConnectionError(Exception):
 
 class SerialConnectionWrapper(object):
 
-    def __init__(self, thread, connection):
+    def __init__(self, thread, connection, serial, protocol_factory):
         self.thread = thread
         self.connection = connection
+        self.serial = serial
+        self.protocol_factory = protocol_factory
 
     def close(self):
         self.thread.close()
 
+    def suspend(self):
+        self.thread.stop()
+        if self.thread.is_alive():
+            raise SerialConnectionError('Failed to suspend serial thread')
+
+    def resume(self):
+        self.thread = ReaderThread(self.serial, self.protocol_factory)
+        self.connection = self.thread.__enter__()
+
+    def is_suspended(self):
+        return False if self.thread.is_alive() else True
+
     def write_line(self, line):
         logger.debug('Writing line to serial port: {line}'.format(line=line))
         self.connection.write_line(line)
 
     def parse_raw_line(self, line):
         self.connection.parse_raw_line(line)
 
+    def instance(self):
+        return self.serial
+
 
 def start_serial_connection(port, baudrate, virtual, timeout, messagebus, entity, filters):
     serial_thread = None
     try:
         serial = Serial(port, baudrate, rtscts=virtual, dsrdtr=virtual, timeout=timeout)
         _lock_serial_port(serial)
-        serial_thread = ReaderThread(serial, _serial_connection(messagebus, entity, filters))
+        protocol_factory = _serial_connection(messagebus, entity, filters)
+        serial_thread = ReaderThread(serial, protocol_factory)
 
-        return SerialConnectionWrapper(serial_thread, serial_thread.__enter__())
+        return SerialConnectionWrapper(
+            serial_thread, serial_thread.__enter__(), serial, protocol_factory)
```

Comment:
Create a ConnectionReaderThreadFactory of a sort that handles the creation and returns the connection and readerthread. It can be done by an internal function or closure that has access to all variables needed.
That way the SerialConnectionWrapper doesn't need to know how to recreate the thread and instead just call the factory. That eliminates the need for all four arguments and replaces it with one - the factory.

Reply:
Agreed. Will fix in the coming days.

Reply:
Fixed now. I'll take a look at the other stuff shortly :smiley: